substratevm/src/com.oracle.svm.core.graal.aarch64/src/com/oracle/svm/core/graal/aarch64/SubstrateAArch64Backend.java

```diff
@@ -171,6 +172,20 @@ public SubstrateAArch64Backend(Providers providers) {
         super(providers);
     }
 
+    static class AArch64AddressFactory {
+        static AArch64Address createWrappedUnscaledImmediateAddress(AArch64MacroAssembler masm, Register base, int imm, int immBits, ScratchRegister scratch) {
```

Comment:
I think we should pass `Register` rather than `ScratchRegister` here. It gives the caller the choice whether a scratch or other register is used here.

Can't we move this static method into `AArch64Address`?

Reply:
Hi. Sorry for the delay - got caught up making sure I could graduate :). Yes, I'll change the ScratchRegister to a Register.

Originally I put `AArch64AddressFactory` in SubstrateAArch64Backend because I didn't want to pollute the non-substrate code. However, It looks like the same method is being called a couple other places too. The use of it in `AArch64HotSpotPatchReturnAddressOp` looks fishy to me. I'll look more into it tomorrow

Reply:
Congratulations on graduating, Tom! I'll leave Stefan to comment on the patch :-)